dali/core/mm/async_pool_test.cu

```diff
@@ -0,0 +1,325 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+#include <random>
+#include <vector>
+#include "dali/core/mm/async_pool.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/mm/mm_test_utils.h"
+#include "dali/core/cuda_stream.h"
+#include "rmm/mr/device/pool_memory_resource.hpp"
+
+namespace dali {
+namespace mm {
+
+struct GPUHog {
+  ~GPUHog() {
+    if (mem) {
+      CUDA_DTOR_CALL(cudaFree(mem));
+      mem = nullptr;
+    }
+  }
+
+  void init() {
+    if (!mem)
+      CUDA_CALL(cudaMalloc(&mem, size));
+  }
+
+  void run(cudaStream_t stream, int count = 1) {
+    for (int i = 0; i < count; i++) {
+      CUDA_CALL(cudaMemsetAsync(mem, i+1, size, stream));
+    }
+  }
+
+  uint8_t *mem = nullptr;
+  size_t size = 16<<20;
+};
+
+TEST(MMAsyncPool, SingleStreamReuse) {
+  GPUHog hog;
+  hog.init();
+
+  CUDAStream stream = CUDAStream::Create(true);
+  test::test_device_resource upstream;
+
+  async_pool_base<memory_kind::device, free_tree, std::mutex> pool(&upstream);
+  stream_view sv(stream);
+  int size1 = 1<<20;
+  void *ptr = pool.allocate_async(size1, sv);
+  hog.run(stream, 2);
+  pool.deallocate_async(ptr, size1, sv);
+  void *p2 = pool.allocate_async(size1, sv);
+  CUDA_CALL(cudaStreamSynchronize(stream));
+  EXPECT_EQ(ptr, p2);
+}
+
+__global__ void Check(const void *ptr, size_t size, uint8_t fill, int *failures) {
+  size_t idx = static_cast<size_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  if (idx < size) {
+    if (static_cast<const uint8_t*>(ptr)[idx] != fill)
+      atomicAdd(failures, 1);
+  }
+}
+
+namespace {
+
+struct block {
+  void *ptr;
+  size_t size;
+  uint8_t fill;
+  cudaStream_t stream;
+};
+
+template <typename Pool, typename Mutex>
+void AsyncPoolTest(Pool &pool, vector<block> &blocks, Mutex &mtx, CUDAStream &stream,
+                   int max_iters = 100000, bool use_hog = false) {
+  stream_view sv(stream);
+  std::mt19937_64 rng(12345);
+  std::poisson_distribution<> size_dist(1024);
+  const int max_size = 1<<20;
+  std::uniform_int_distribution<> sync_dist(10, 10);
+  std::bernoulli_distribution action_dist;
+  std::bernoulli_distribution hog_dist(0.05f);
+  std::uniform_int_distribution<> fill_dist(1, 255);
+  DeviceBuffer<int> failure_buf;
+  int failures = 0;
+  failure_buf.from_host(&failures, 1, stream);
+  GPUHog hog;
+  if (use_hog)
+    hog.init();
+  int hogs = 0;
+  int max_hogs = sync_dist(rng);
+  CUDAEvent event = CUDAEvent::Create();
+  for (int i = 0; i < max_iters; i++) {
+    if (use_hog && hog_dist(rng)) {
+      if (hogs++ > max_hogs) {
```

Comment:
So the bigger the `hogs` value is the less likely the code here is happen to be executed, right?

Reply:
max_hogs are max scheduled hogs - when we've scheduled max_hogs, a synchronization is done to avoid scheduling too much work ahead.